File: src/updater.py
"""
自动更新检查 - 查询 GitHub Releases 最新版本

提供非阻塞的更新检查：
- 在后台线程中请求 GitHub API
- 解析语义化版本号并比较
- 返回新版本信息或错误说明
"""
import re
import threading
from typing import NamedTuple, Optional

import requests

from logger import logger


class UpdateInfo(NamedTuple):
    """更新检查结果"""
    has_update: bool
    current_version: str
    latest_version: str
    release_url: str
    release_notes: str
    error: str = ""


GITHUB_RELEASES_API = "https://api.github.com/repos/{owner}/{repo}/releases/latest"


def parse_version(version: str) -> tuple[int, ...]:
    """将版本字符串解析为可比较的整数元组，忽略 v 前缀"""
    cleaned = version.lower().lstrip("v").strip()
    parts = re.split(r"[.-]", cleaned)
    numbers = []
    for part in parts:
        try:
            numbers.append(int(part))
        except ValueError:
            break
    return tuple(numbers) if numbers else (0,)


def is_newer(current: str, latest: str) -> bool:
    """判断 latest 是否比 current 新"""
    return parse_version(latest) > parse_version(current)


def check_update(
    current_version: str,
    owner: str = "Qiuini",
    repo: str = "BiliHistory",
    timeout: int = 10,
) -> UpdateInfo:
    """同步检查 GitHub Releases 是否有新版本"""
    url = GITHUB_RELEASES_API.format(owner=owner, repo=repo)
    try:
        resp = requests.get(
            url,
            timeout=timeout,
            headers={
                "Accept": "application/vnd.github+json",
                "User-Agent": f"BiliHistory/{current_version}",
                "X-GitHub-Api-Version": "2022-11-28",
            },
        )
        resp.raise_for_status()
        data = resp.json()

        latest = data.get("tag_name", "").strip()
        release_url = data.get("html_url", "")
        release_notes = data.get("body", "") or ""

        if not latest:
            return UpdateInfo(
                has_update=False,
                current_version=current_version,
                latest_version="",
                release_url="",
                release_notes="",
                error="无法解析最新版本号",
            )

        has_update = is_newer(current_version, latest)
        return UpdateInfo(
            has_update=has_update,
            current_version=current_version,
            latest_version=latest,
            release_url=release_url,
            release_notes=release_notes,
        )

    except requests.exceptions.HTTPError as e:
        status = e.response.status_code if e.response else 0
        if status == 404:
            return UpdateInfo(
                False, current_version, "", "", "", "未找到发布页面（仓库可能没有 Release）"
            )
        return UpdateInfo(False, current_version, "", "", "", f"GitHub API 错误 ({status})")
    except requests.exceptions.Timeout:
        return UpdateInfo(False, current_version, "", "", "", "检查更新超时")
    except requests.exceptions.RequestException as e:
        return UpdateInfo(False, current_version, "", "", "", f"网络请求失败: {e}")
    except Exception as e:
        return UpdateInfo(False, current_version, "", "", "", f"检查更新异常: {e}")


class UpdateChecker(threading.Thread):
    """后台更新检查线程"""

    def __init__(
        self,
        current_version: str,
        owner: str = "Qiuini",
        repo: str = "BiliHistory",
        callback=None,
    ):
        super().__init__(daemon=True)
        self.current_version = current_version
        self.owner = owner
        self.repo = repo
        self.callback = callback

    def run(self):
        info = check_update(self.current_version, self.owner, self.repo)
        if self.callback:
            try:
                self.callback(info)
            except Exception as e:
                logger.warning(f"更新检查回调异常: {e}")

File: src/business/updater.cpp
#include "updater.h"

#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonParseError>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QTimer>
#include <QUrl>

namespace bili::business {

namespace {
// 更新检查总超时：覆盖 DNS/连接/传输全程，避免弱网下 helper 永久挂起泄漏。
constexpr int kUpdateCheckTimeoutMs = 15000;
} // namespace

QList<int> parseVersion(const QString& version)
{
    QList<int> result;
    QString trimmed = version;
    if (trimmed.startsWith('v', Qt::CaseInsensitive)) {
        trimmed = trimmed.mid(1);
    }

    const QStringList parts = trimmed.split('.');
    for (const QString& part : parts) {
        bool ok = false;
        const int value = part.toInt(&ok);
        result.append(ok ? value : 0);
    }

    while (result.size() < 3) {
        result.append(0);
    }
    return result;
}

bool isNewer(const QString& current, const QString& latest)
{
    const QList<int> currentParts = parseVersion(current);
    const QList<int> latestParts = parseVersion(latest);

    const int maxLen = qMax(currentParts.size(), latestParts.size());
    for (int i = 0; i < maxLen; ++i) {
        const int c = i < currentParts.size() ? currentParts[i] : 0;
        const int l = i < latestParts.size() ? latestParts[i] : 0;
        if (l != c) {
            return l > c;
        }
    }
    return false;
}

void checkUpdate(const QString& currentVersion,
                 const QString& owner,
                 const QString& repo,
                 std::function<void(UpdateInfo)> callback)
{
    if (!callback) {
        return;
    }

    auto* helper = new QObject();
    auto* manager = new QNetworkAccessManager(helper);

    const QUrl url(QStringLiteral("https://api.github.com/repos/%1/%2/releases/latest")
                       .arg(owner, repo));
    QNetworkRequest request(url);
    request.setRawHeader("Accept", "application/vnd.github+json");
    request.setRawHeader("User-Agent",
                         QStringLiteral("BiliHistory/%1").arg(currentVersion).toUtf8());
    request.setRawHeader("X-GitHub-Api-Version", "2022-11-28");
    request.setTransferTimeout(kUpdateCheckTimeoutMs);

    QNetworkReply* reply = manager->get(request);
    // 看门狗：transferTimeout 仅覆盖传输阶段，DNS/连接挂起时需兜底强制中止并清理。
    QTimer::singleShot(kUpdateCheckTimeoutMs, helper, [helper, reply]() {
        if (reply) reply->abort();
        helper->deleteLater();
    });
    QObject::connect(reply, &QNetworkReply::finished, helper,
                     [helper, reply, currentVersion, callback]() {
        UpdateInfo info;
        info.currentVersion = currentVersion;

        if (reply->error() != QNetworkReply::NoError) {
            info.error = QStringLiteral("网络请求失败: %1").arg(reply->errorString());
            callback(info);
            helper->deleteLater();
            return;
        }

        const QByteArray data = reply->readAll();
        QJsonParseError parseError;
        const QJsonDocument doc = QJsonDocument::fromJson(data, &parseError);
        if (parseError.error != QJsonParseError::NoError || !doc.isObject()) {
            info.error = QStringLiteral("解析响应失败: %1").arg(parseError.errorString());
            callback(info);
            helper->deleteLater();
            return;
        }

        const QJsonObject obj = doc.object();
        info.latestVersion = obj.value(QStringLiteral("tag_name")).toString();
        info.releaseUrl = obj.value(QStringLiteral("html_url")).toString();
        info.releaseNotes = obj.value(QStringLiteral("body")).toString();

        if (info.latestVersion.isEmpty()) {
            info.error = QStringLiteral("响应中未找到版本号");
            callback(info);
            helper->deleteLater();
            return;
        }

        info.hasUpdate = isNewer(currentVersion, info.latestVersion);
        callback(info);
        helper->deleteLater();
    });
}

} // namespace bili::business
